File: apps/mobile-app/ios/VaultStoreKit/Database/Repositories/ItemStatsRepository.swift
import Foundation
import VaultModels

/// The actions whose use of an item is recorded. Each maps to its own timestamp + counter pair alongside the
/// aggregate `LastUsedAt` / `UseCount`.
public enum ItemUsageAction {
    /// A credential filled into a form.
    case autofill
    /// A value copied to the clipboard.
    case copy
    /// A passkey assertion.
    case passkey

    /// The column pair this action bumps. Closed set: nothing here is ever built from caller input.
    internal var columns: (last: String, count: String) {
        switch self {
        case .autofill:
            return ("LastAutofilledAt", "AutofillCount")
        case .copy:
            return ("LastCopiedAt", "CopyCount")
        case .passkey:
            return ("LastPasskeyAuthAt", "PasskeyAuthCount")
        }
    }
}

/// Repository for per-item usage statistics.
public class ItemStatsRepository: BaseRepository {
    /// Record one use of an item. Runs in its own transaction so the vault is persisted and marked dirty.
    /// - Parameters:
    ///   - itemId: The item that was used
    ///   - manifestId: The manifest the item belongs to
    ///   - action: What the user did with it
    public func recordUsage(itemId: String, manifestId: String, action: ItemUsageAction) throws {
        // Usage statistics live in their own data bucket which should be pushed without a full manifest write.
        try withTransaction(scope: VaultDataBucketCategory.stats) {
            let now = self.now()
            let columns = action.columns
            try client.executeUpdate(ItemStatsQueries.insertRow, params: [manifestId, itemId, now, now])
            try client.executeUpdate(ItemStatsQueries.forAction(lastColumn: columns.last, countColumn: columns.count), params: [now, now, now, manifestId, itemId])
        }
    }
}

File: apps/mobile-app/android/app/src/main/java/net/aliasvault/app/vaultstore/repositories/ItemStatsRepository.kt
package net.aliasvault.app.vaultstore.repositories

import net.aliasvault.app.vaultstore.VaultDatabase
import net.aliasvault.app.vaultstore.models.VaultDataBucketCategory
import net.aliasvault.app.vaultstore.queries.ItemStatsQueries

/**
 * The actions whose use of an item is recorded, each with its own timestamp + counter pair.
 */
enum class ItemUsageAction(
    /** The per-action timestamp column. */
    val lastColumn: String,
    /** The per-action counter column. */
    val countColumn: String,
) {
    /** The item was filled into a form. */
    AUTOFILL("LastAutofilledAt", "AutofillCount"),

    /** A value of the item was copied. */
    COPY("LastCopiedAt", "CopyCount"),

    /** The item's passkey signed an assertion. */
    PASSKEY("LastPasskeyAuthAt", "PasskeyAuthCount"),
}

/**
 * Repository for per-item usage statistics.
 */
class ItemStatsRepository(database: VaultDatabase) : BaseRepository(database) {
    /**
     * Record one use of an item, in its own transaction so the vault is persisted and marked dirty.
     * @param itemId The item that was used
     * @param manifestId The manifest the item belongs to
     * @param action What the user did with it
     */
    fun recordUsage(itemId: String, manifestId: String, action: ItemUsageAction) {
        // Usage statistics live in their own data bucket which should be pushed without a full manifest write.
        withTransaction(VaultDataBucketCategory.STATS) {
            val now = now()
            val id = itemId.lowercase()
            executeUpdate(ItemStatsQueries.INSERT_ROW, arrayOf(manifestId, id, now, now))
            executeUpdate(ItemStatsQueries.forAction(action.lastColumn, action.countColumn), arrayOf(now, now, now, manifestId, id))
        }
    }
}
